ipwb/replay.py

```diff
@@ -179,6 +179,126 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def lookup_keys(surt):
+    keyre = re.compile(r"(.+)([,/]).+")
+    key = surt.split("?")[0].strip("/")
+    keys = [key, f"{key}/*"]
+    while "," in key:
+        try:
+            m = keyre.match(key)
+            keys.append(f"{m[1]}{m[2]}*")
+            key = m[1]
+        except Exception as e:
+            key = key.strip("/,")
+
+    return keys
+
+
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, rest = ln.split(maxsplit=1)
+    if key == surtk:
+        return [ln]
+
+    # If further searching required...
+    left = 0
+    # Go to end of seek stream
+    iter.seek(0, 2)
+    right = iter.tell()  # Current position
+
+    lines = set()  # Prevents dupes
+    while (right - left > 1):
+        mid = (right + left) // 2
+        iter.seek(mid)
+        iter.readline()  # Purge rest of current line
+        line = iter.readline()  # Read the next full line
+
+        if len(line) == 0:
+            right = mid
+            continue
+
+        surtk, rest = line.split(maxsplit=1)
+        datetimeK = rest.split()[0].decode()
+        if surtk[-1:] == b'/':
+            surtk = surtk[0:-1]
+
+        # TODO: find a more elegant way for comparison than manually
+        # trimming the last two chars off of the surtk
+
+        if key == surtk:
+            if datetime and datetime != datetimeK:
+                # Rm other close matches, exact found
+                lines.clear()
+                lines.add(line)
+
+                break
+            lines.add(line)
+            # Iterate further to get lines after selection point
+            nextLine = iter.readline()
+            while nextLine:
+                surtk, rest = nextLine.split(maxsplit=1)
+                datetimeK = rest.split()[0].decode()
+                if surtk[-1:] == b'/':
+                    surtk = surtk[0:-1]
+
+                if key == surtk:
+                    lines.add(nextLine)
+
+                nextLine = iter.readline()
+
+            # Continue searching until find first instance
+            right = mid
+        elif key > surtk:
+            left = mid
+        else:
+            right = mid
+
+    # Convert uniq set to list then sort and return
+    ret = sorted(list(lines))
+
+    return ret
+
+
+def run_batchlookup(surt, filename, datetime=None):
+    mobj = open(filename, "rb")
+    iter = mobj
+
+    fobj = open(filename, "rb")
+    for line in fobj:
+        res = lookup(iter, surt, datetime)
+        if res:
+            return res
+    fobj.close()
+    mobj.close()
+
+
+def lookup(iter, surt, datetime=None):
+    for idx, key in enumerate(lookup_keys(surt)):
+        res = bin_search(iter, key.encode(), datetime)
+        if res:
+            return res
+
+
```

Comment:
What are these functions doing here? These were specific to MementoMap where I had to find partial key matches.

```suggestion
```

Reply:
Coded in similar paths through the code. Will simplify calls and remove.

Reply:
In MementoMap, I did it this way because the need there was to find the longest common sub-URI-key. So, for a given surt `foo,bar)/baz/blah`, I was supposed to check for any matches in the following order and terminate as soon as there is a match:

```
foo,bar)/baz/blah
foo,bar)/baz/blah/*
foo,bar)/baz/*
foo,bar)/*
foo,bar,*
foo,*
```

In this case, `lookup_keys()` was returning such a list of partial keys that was then binary searched.